processing/context_collector.go

```diff
@@ -0,0 +1,198 @@
+// alert.
+type ContextShipper func(Context, *log.Entry) error
+
+// ContextCollectorPerfStats contains performance stats written to InfluxDB
+// for monitoring.
+type ContextCollectorPerfStats struct {
+	Flows     uint64 `influx:"context_flows"`
+	Events    uint64 `influx:"context_events"`
+	JSONBytes uint64 `influx:"context_json_bytes"`
+}
+
+// ContextCollector is a component that maintains a cache of metadata per
+// flow ID, forwarding it to a specified sink if associated with an alert.
+type ContextCollector struct {
+	PerfStats          ContextCollectorPerfStats
+	StatsEncoder       *util.PerformanceStatsEncoder
+	StopChan           chan bool
+	StoppedChan        chan bool
+	StopCounterChan    chan bool
+	StoppedCounterChan chan bool
+	Running            bool
+	StatsLock          sync.Mutex
+
+	Cache    *cache.Cache
+	MarkLock sync.Mutex
+	Marked   map[string]struct{}
+	Logger   *log.Entry
+	i        uint64
+	Ship     ContextShipper
+}
+
+// Context is a collection of JSON events that belong to a given flow.
+type Context []string
+
+// MakeContextCollector creates a new ContextCollector.
+func MakeContextCollector(shipper ContextShipper) *ContextCollector {
+	c := &ContextCollector{
+		Logger: log.WithFields(log.Fields{
+			"domain": "context",
+		}),
+		Cache:  cache.New(ContextCollectorDefaultTTL, ContextCollectorDefaultTTL),
+		Marked: make(map[string]struct{}),
+		i:      0,
+		Ship:   shipper,
+	}
+	return c
+}
+
+// Mark queues metadata for a given flow for forwarding, identified by its
+// flow ID.
+func (c *ContextCollector) Mark(flowID string) {
+	// when seeing an alert, just mark the flow ID as relevant
+	c.MarkLock.Lock()
+	c.Marked[flowID] = struct{}{}
+	c.MarkLock.Unlock()
+}
+
+// Consume processes an Entry, adding the data within to the internal
+// aggregated state
+func (c *ContextCollector) Consume(e *types.Entry) error {
+	var myC Context
+	// Some events, e.g. stats, have no flow ID set
+	if e.FlowID == "" {
+		return nil
+	}
+
+	cval, exist := c.Cache.Get(e.FlowID)
+	if exist {
+		// the 'flow' event always comes last, so we can use it as an
+		// indicator that the flow is complete and can be processed
+		if e.EventType == "flow" {
+			var isMarked bool
+			c.MarkLock.Lock()
+			if _, ok := c.Marked[e.FlowID]; ok {
+				isMarked = true
+			}
+			c.MarkLock.Unlock()
+			if isMarked {
+				c.StatsLock.Lock()
+				c.PerfStats.Flows++
+				c.PerfStats.Events += uint64(len(cval.(Context)))
+				for _, v := range cval.(Context) {
+					c.PerfStats.JSONBytes += uint64(len(v))
+				}
+				c.StatsLock.Unlock()
+				c.Ship(cval.(Context), c.Logger)
+				delete(c.Marked, e.FlowID)
+			}
+			c.Cache.Delete(e.FlowID)
+		} else {
+			myC = cval.(Context)
+			myC = append(myC, e.JSONLine)
+			c.Cache.Set(e.FlowID, myC, cache.DefaultExpiration)
+		}
+	} else {
+		if e.EventType != "flow" {
+			myC = append(myC, e.JSONLine)
+			c.Cache.Set(e.FlowID, myC, cache.DefaultExpiration)
+		}
+	}
+	c.i++
+	if c.i%100000 == 0 {
```

Comment:
Magic number, and referenced in text of log message: candidate for constant.

Reply:
Done